server/blob/blob-cloud/src/main/java/org/apache/james/blob/cloud/CloudBlobsDAO.java

```diff
@@ -0,0 +1,120 @@
+import com.google.common.hash.Hashing;
+import com.google.common.hash.HashingInputStream;
+import com.google.inject.Inject;
+import com.google.inject.Module;
+
+class CloudBlobsDAO implements ObjectStore {
+    private static final InputStream EMPTY_STREAM = new ByteArrayInputStream(new byte[0]);
+    private static final Iterable<Module> JCLOUDS_MODULES = ImmutableSet.of(new SLF4JLoggingModule());
+
+    private final BlobId.Factory blobIdFactory;
+    private final BlobStore blobStore;
+    private final ContainerName containerName;
+
+    @Inject
+    public CloudBlobsDAO(ContainerName containerName, HashBlobId.Factory blobIdFactory,
+                         CloudsBlobsConfiguration cloudsBlobsConfiguration) {
+        this.blobIdFactory = blobIdFactory;
+        this.containerName = containerName;
+
+        RegionScopedBlobStoreContext blobStoreContext = ContextBuilder.newBuilder("openstack-swift")
+            .endpoint(cloudsBlobsConfiguration.getEndpoint().toString())
+            .credentials(
+                cloudsBlobsConfiguration.getIdentity().getValue(),
+                cloudsBlobsConfiguration.getCredentials().getValue())
+            .overrides(cloudsBlobsConfiguration.getOverrides())
+            .modules(JCLOUDS_MODULES)
+            .buildView(RegionScopedBlobStoreContext.class);
+
+        blobStore = cloudsBlobsConfiguration
+            .getRegion()
+            .map(region -> blobStoreContext.getBlobStore(region.getValue()))
+            .orElse(blobStoreContext.getBlobStore());
+    }
+
+    @Override
+    public CompletableFuture<BlobId> save(byte[] data) {
+        return save(new ByteArrayInputStream(data));
+    }
+
+    @Override
+    public CompletableFuture<BlobId> save(InputStream data) {
+        Preconditions.checkNotNull(data);
+        HashingInputStream hashingInputStream = new HashingInputStream(Hashing.sha256(), data);
+        String tmpId = UUID.randomUUID().toString();
+        String containerName = this.containerName.getValue();
+
+        Blob blob = blobStore.blobBuilder(tmpId).payload(hashingInputStream).build();
+        blobStore.putBlob(containerName, blob);
+        BlobId id = blobIdFactory.from(hashingInputStream.hash().toString());
+        blobStore.copyBlob(containerName, tmpId, containerName, id.asString(), CopyOptions.NONE);
+        blobStore.removeBlob(containerName, tmpId);
+
+        return CompletableFuture.completedFuture(id);
+    }
+
+    @Override
+    public CompletableFuture<byte[]> readBytes(BlobId blobId) {
+        return CompletableFuture
+            .supplyAsync(Throwing.supplier(() -> IOUtils.toByteArray(read(blobId))).sneakyThrow());
+    }
+
+    @Override
+    public InputStream read(BlobId blobId) {
+        Blob blob = blobStore.getBlob(containerName.getValue(), blobId.asString());
+
+        InputStream is = EMPTY_STREAM;
+        try {
+            if (blob != null) {
+                is = blob.getPayload().openStream();
```

Comment:
we usually return as soon as possible and avoid variable redefinition

Reply:
That forces me to have 3 return statements in the method (because of the conditional and the exception), is that fine ?

Reply:
I would just `return EMPTY_STREAM;` at method's end and `return blob.getPayload().openStream();` in the working case